### PR TITLE
Changes to avoid nvram limitations

### PR DESCRIPTION
**Pull request by xchapron-ledger**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -144,6 +144,23 @@ DEFINES += HAVE_UX_STACK_INIT_KEEP_TICKER
...
+DEFINES += HAVE_NO_RESET_GENERATION_INCREMENT
+
+# Disable by default rk support and expose a setting to enable it
+# This means that by default user won't be able to create "Resident Keys",
+# which are also named "Discoverable Credentials".
+# This has been implemented to protect user from the nvram wipe mostly happening
+# during an app update which will erase their RK credentials we no possibility
+# to restore them.
+# Advanced user can still choose to enable this setting at their own risk.
+DEFINES += HAVE_RK_SUPPORT_SETTING
```
Comment (fbeutin-ledger) [inline Makefile:162]: Why is a define used ? if we still have a boolean in the settings behind to active deactivate the feature ?

File: include/config.h [hunk 2]
```diff
@@ -47,4 +52,8 @@ void config_set_ctap2_pin(uint8_t *pin);
 void config_decrease_ctap2_pin_retry_counter(void);
 void config_reset_ctap2_pin_retry_counter(void);
 
+#ifdef HAVE_RK_SUPPORT_SETTING
+void config_set_rk_enabled(uint8_t enabled);
```
Comment (fbeutin-ledger) [inline include/config.h:56]: `config_set_rk_status(bool enabled)`

File: Makefile (hunk 1 above)
Comment (fbeutin-ledger) [reply Makefile:162]: Quite confusing, at first I wondered why we say it's disabled by default yet add the flag in the define

File: src/ui_shared.c [hunk 3]
```diff
@@ -18,6 +18,113 @@
...
+        &ux_settings_warning_flow_2_step,
+        &ux_settings_warning_flow_3_step,
+        FLOW_LOOP);
+
+static void display_warning() {
+    ux_flow_init(0, ux_settings_warning_flow, NULL);
+}
+
+static void toogle_settings() {
+    if (N_u2f.rk_enabled) {
```
Comment (fbeutin-ledger) [inline src/ui_shared.c:65]: config_set_rk_enabled(!N_u2f.rk_enabled)

File: src/ui_shared.c [hunk 4]
```diff
@@ -18,6 +18,113 @@
...
+                         "credentials will be lost.",
+             });
+
+UX_STEP_CB(ux_settings_warning_flow_2_step,
+           pbn,
+           toogle_settings(),
+           {
+               &C_icon_validate_14,
+               "Enable",
+               "Discoverable Creds",
```
Comment (fbeutin-ledger) [inline src/ui_shared.c:43]: `Discoverable credentials` ?

File: src/ui_shared.c [hunk 5]
```diff
@@ -18,6 +18,113 @@
...
+
+UX_STEP_CB(ux_settings_flow_1_disabled_step,
+           nb_paging,
+           display_warning(),
+           {"Discoverable Credentials", "Disabled"});
+
+UX_STEP_NOCB(ux_settings_flow_2_disabled_step,
+             nn_paging,
+             {
+                 .title = "Discoverable Creds",
```
Comment (fbeutin-ledger) [inline src/ui_shared.c:81]: `Discoverable credentials` ?

File: src/ui_shared.c [hunk 6]
```diff
@@ -18,6 +18,113 @@
...
+
+UX_STEP_CB(ux_settings_flow_1_enabled_step,
+           nb_paging,
+           toogle_settings(),
+           {"Discoverable Credentials", "Enabled"});
+
+UX_STEP_NOCB(ux_settings_flow_2_enabled_step,
+             nn_paging,
+             {
+                 .title = "Discoverable Creds",
```
Comment (fbeutin-ledger) [inline src/ui_shared.c:94]: Why just `Creds` ? and why a capital ?

Comment (xchapron-ledger) [reply src/ui_shared.c:94]: @fbeutin-ledger `Discoverable Credentials` don't fit in the screen when in bold...
Will ask the product what they think about it.

File: Makefile (hunk 1 above)
Comment (xchapron-ledger) [reply Makefile:162]: The define is here to easily allow to remove the setting and restore the previously default behavior where RK option support was always enabled.

Comment (fbeutin-ledger) [reply Makefile:162]: Ok

**Comment by fbeutin-ledger** [comment]
Also it seems that we have issues on clang analyzer coming from the SDK ...

**Comment by xchapron-ledger** [comment]
> Also it seems that we have issues on clang analyzer coming from the SDK ...

Yep, this is specific to FIDO2 protocol, and this has been fixed on API_LEVEL >= 2 but never cherry-picked..
I think this is not a big deal as it impact only this app. I'd say lets wait until it disappear by itself with next Os release :)

File: src/ui_shared.c (hunk 6 above)
Comment (xchapron-ledger) [reply src/ui_shared.c:94]: "Resident key" is the technical naming. Users should normally see "Discoverable Credentials" in classical flow.

File: src/config.c [hunk 7]
```diff
@@ -82,6 +82,12 @@ void config_init(void) {
         tmp8 = 0;
         nvm_write((void *) &N_u2f.pinSet, (void *) &tmp8, sizeof(uint8_t));
 
+#ifdef HAVE_RK_SUPPORT_SETTING
+        // Initialize rk_enable value: Disabled by default
+        tmp8 = 0;
```
Comment (fbeutin-ledger) [inline src/config.c:87]: Use a DEFINED to avoid manipulating literals
```
RK_ENABLED
RK_DISABLED
```

File: src/config.c [hunk 8]
```diff
@@ -104,14 +110,16 @@ uint8_t config_increase_and_get_authentification_counter(uint8_t *buffer) {
...
     uint32_t resetGeneration = N_u2f.resetGeneration + 1;
-    uint8_t pinSet = 0;
 
     nvm_write((void *) &N_u2f.resetGeneration, (void *) &resetGeneration, sizeof(uint32_t));
 
     // Update keys derived from seed
     derive_and_store_keys(N_u2f.resetGeneration);
+#endif
 
+    uint8_t pinSet = 0;
```
Comment (fbeutin-ledger) [inline src/config.c:122]: same for pinset

**Review by fbeutin-ledger — APPROVED** [review]
Approved with optional small change request

**Pull request merged**